onnx2kerastl/reshape_layers.py

```diff
@@ -65,6 +66,45 @@ def convert_shape(node, params, layers, lambda_func, node_name, keras_name):
         layers[node_name] = tf.shape(input_0, out_type=tf.int64)
 
 
+def optimize_constant_array_for_serialization(params: tf.Tensor, indices: Union[np.ndarray, tf.Tensor], logger):
+    """
+    Some models contain repetition in the constant array tf.gather gathers from.
+    In such cases, serialization takes a long time. serializing 3000 elements take ~3 sec (and scale linearly).
+    To optimize for this case, we detect the repetition, take the modulu of the repetition from indices,
+    and making the param array shorter.
+    This is especially useful for models containing pre-top-k elements.
+    Args:
+        params: the array to gather on
+        indices: the indices to gather
+
+    Returns:
+        params: the new array to gather on
+        indices: the new indices to gather
+    """
+    # This Optimization is a must for models with Pre-Top-K needs
+    logger.debug('onnx2keras.gather - Encountered long gather. '
+                 'Trying to shorten it for easy serialization')
+    max_inp = tf.reduce_max(params)
+    min_inp = tf.reduce_min(params)
+    if len(params) % (max_inp - min_inp + 1) == 0:
+        range_inp = tf.range(min_inp, max_inp + 1, dtype=params.dtype)
+        reshape_columns = tf.reshape(params, [-1, max_inp + 1])
+        if tf.reduce_sum(tf.abs((reshape_columns - range_inp[None, :]))) == 0:
+            # [0,1,2,3,0,1,2,3]
+            logger.debug('onnx2keras.gather - Shortening sequence - columns')
+            indices = indices % (max_inp + 1)
+            params = range_inp
+        else:
+            repetition_len = np.argmin(params == params[0])
+            if repetition_len > 0 and len(params) % repetition_len == 0:
+                reshaped_rows = tf.reshape(params, [-1, repetition_len])
+                first_row = reshaped_rows[:, 0]
+                if tf.reduce_sum(tf.abs(reshaped_rows - first_row[:, None])) == 0:
```

Comment:
Maybe add a line comment for each if/else to explain in plain words what is it checking

Reply:
Done